processor/src/trace/virtual_bus/sum_check/prover/mod.rs

```diff
@@ -0,0 +1,323 @@
+            round_proofs,
+        })
+    }
+
+    /// Proves a round of the sum-check protocol.
+    pub fn prove_rounds(
+        &self,
+        claim: E,
+        mls: &mut [MultiLinearPoly<E>],
+        num_rounds: usize,
+        coin: &mut C,
+    ) -> Result<(RoundClaim<E>, Vec<RoundProof<E>>), Error> {
+        // there should be at least one multi-linear polynomial provided
+        if mls.is_empty() {
+            return Err(Error::NoMlsProvided);
+        }
+
+        // there should be at least one round to prove
+        if num_rounds == 0 {
+            return Err(Error::NumRoundsZero);
+        }
+
+        // there can not be more rounds than variables of the multi-linears
+        let ml_variables = mls[0].num_variables();
+        if num_rounds > ml_variables {
+            return Err(Error::TooManyRounds);
+        }
+
+        // there should at least be one variable for the protocol to be non-trivial
+        if ml_variables < 1 {
+            return Err(Error::AtLeastOneVariable);
+        }
+
+        // all multi-linears should have the same arity
+        if !mls.iter().all(|ml| ml.num_variables() == ml_variables) {
+            return Err(Error::MlesDifferentArities);
+        }
+
+        let mut round_proofs = vec![];
+
+        // setup first round claim
+        let mut current_round_claim = RoundClaim {
+            eval_point: vec![],
+            claim,
+        };
+
+        // run the first round of the protocol
+        let round_poly_evals = sumcheck_round(&self.composition_poly, mls);
+        let round_poly_coefs = round_poly_evals.to_poly(current_round_claim.claim);
+
+        // reseed with the s_0 polynomial
+        coin.reseed(H::hash_elements(&round_poly_coefs.coefficients));
+        round_proofs.push(RoundProof { round_poly_coefs });
+
+        for i in 1..num_rounds {
+            // generate random challenge r_i for the i-th round
+            let round_challenge = coin.draw().map_err(|_| Error::FailedToGenerateChallenge)?;
+
+            // compute the new reduced round claim
+            let new_round_claim =
+                reduce_claim(&round_proofs[i - 1], current_round_claim, round_challenge);
+
+            // fold each multi-linear using the round challenge
+            mls.iter_mut().for_each(|ml| ml.bind(round_challenge));
+
+            // run the i-th round of the protocol using the folded multi-linears for the new reduced
+            // claim. This basically computes the s_i polynomial.
+            let round_poly_evals = sumcheck_round(&self.composition_poly, mls);
+
+            // update the claim
+            current_round_claim = new_round_claim;
+
+            let round_poly_coefs = round_poly_evals.to_poly(current_round_claim.claim);
+
+            // reseed with the s_i polynomial
+            coin.reseed(H::hash_elements(&round_poly_coefs.coefficients));
+            let round_proof = RoundProof { round_poly_coefs };
+            round_proofs.push(round_proof);
+        }
```

Comment:
Why do we introduce `new_round_claim` variable here? Seems like the following should be equivalent:
```Rust
for i in 1..num_rounds {
    // generate random challenge r_i for the i-th round
    let round_challenge = coin.draw().map_err(|_| Error::FailedToGenerateChallenge)?;

    // compute the new reduced round claim
    current_round_claim =
        reduce_claim(&round_proofs[i - 1], current_round_claim, round_challenge);

    // fold each multi-linear using the round challenge
    mls.iter_mut().for_each(|ml| ml.bind(round_challenge));

    // run the i-th round of the protocol using the folded multi-linears for the new reduced
    // claim. This basically computes the s_i polynomial.
    let round_poly_evals = sumcheck_round(&self.composition_poly, mls);
    let round_poly_coefs = round_poly_evals.to_poly(current_round_claim.claim);

    // reseed with the s_i polynomial
    coin.reseed(H::hash_elements(&round_poly_coefs.coefficients));
    let round_proof = RoundProof { round_poly_coefs };
    round_proofs.push(round_proof);
}
```

Reply:
That's mainly for readability. No strong preference from my side though